compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1835,6 +1836,75 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    pub(super) fn collect_unused_for_coerce_return_ty(
+        &self,
+        errors: &Vec<traits::FulfillmentError<'tcx>>,
+    ) {
+        struct SemiVisitor<'a, 'tcx> {
+            fcx: &'a FnCtxt<'a, 'tcx>,
+            semi_expr: Option<&'tcx hir::Expr<'tcx>>,
+            fail_span: Span,
+            return_ty: Ty<'tcx>,
+        }
+
+        impl<'a, 'tcx> Visitor<'tcx> for SemiVisitor<'a, 'tcx> {
+            fn visit_stmt(&mut self, s: &'tcx Stmt<'tcx>) {
+                match s.kind {
+                    hir::StmtKind::Semi(ref expr) => {
+                        if !matches!(expr.kind, hir::ExprKind::Ret(..))
+                            && expr.span.contains(self.fail_span)
+                        {
+                            let expr_ty = self.fcx.typeck_results.borrow().expr_ty(expr);
+                            let expr_ty = self.fcx.tcx.erase_regions(expr_ty);
+                            if self.fcx.can_coerce(expr_ty, self.return_ty) {
+                                self.semi_expr = Some(expr);
+                            }
+                        }
+                    }
+                    _ => {}
+                }
+                intravisit::walk_stmt(self, s);
+            }
+        }
+
+        for fillment_error in errors.iter() {
+            let obligation = &fillment_error.obligation;
+            let body_id = obligation.cause.body_id;
+            let span = obligation.cause.span;
+            let Some(mut diag) =
+                self.tcx.sess.diagnostic().steal_diagnostic(span, StashKey::MaybeForgetReturn)
+            else {
+                continue;
+            };
+
+            if let Some(body_id) = self.tcx.hir().maybe_body_owned_by(
+                self.tcx.typeck_root_def_id(body_id.to_def_id()).expect_local(),
+            ) {
+                let body = self.tcx.hir().body(body_id).value;
+                if let Some((fn_sig_hir_id, fn_decl, _)) = self.get_fn_decl(body.hir_id)
+                    && let hir::FnRetTy::Return(ty) = fn_decl.output {
+                        let ty = self.astconv().ast_ty_to_ty(ty);
+                        let bound_vars = self.tcx.late_bound_vars(fn_sig_hir_id);
+                        let ty = self.tcx.erase_late_bound_regions(Binder::bind_with_vars(ty, bound_vars));
```

Comment:
Can you use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir_typeck/struct.FnCtxt.html#method.body_fn_sig ?